Add Heartbeat properties for App Services

d3r3kk opened the pull request:
Fix Issue #868.
Add heartbeat properties when App Services environment is detected. Keep App Services environment in sync using existing telemetry initializer used for same purpose.

- [x] I ran Unit Tests locally.
- [x] Changes in public surface reviewed
   - No change to public API
- [x] Design discussion issue #868
- [x] CHANGELOG.md updated with one line description of the fix, and a link to the original issue.

d3r3kk commented:
@SergeyKanzhelev can you please have a look at this and let me know if this satisfies the topics we covered in issue #868? I believe I captured all the items discussed there, but I do still need to find a way to allow for updates in non-internal scenarios (AspNetCore).

SergeyKanzhelev commented on Src/WindowsServer/WindowsServer.Shared/AppServicesHeartbeatTelemetryModule.cs:
```diff
@@ -0,0 +1,150 @@
+    /// Provides default values for the heartbeat feature of Application Insights that
+    /// are specific to Azure App Services (Web Apps, Functions, etc...).
+    /// </summary>
+    internal class AppServicesHeartbeatTelemetryModule : ITelemetryModule, IDisposable
```
must be public so it can be instantiated from code

d3r3kk replied:
Done

SergeyKanzhelev commented on Src/WindowsServer/WindowsServer.Shared/AppServicesHeartbeatTelemetryModule.cs:
```diff
@@ -0,0 +1,150 @@
+        /// <summary>
+        /// Initializes a new instance of the<see cref="AppServicesHeartbeatTelemetryModule" /> class.
+        /// </summary>
+        internal AppServicesHeartbeatTelemetryModule()
```
must be public so it can be instantiated from code

d3r3kk replied:
Done. Totally missed this one. Thanks.

SergeyKanzhelev commented on Src/WindowsServer/WindowsServer.Shared/AppServicesHeartbeatTelemetryModule.cs:
```diff
@@ -0,0 +1,150 @@
+
+        private static AppServicesHeartbeatTelemetryModule instance;
+        private object lockObject = new object();
+        private bool isInitialized = false;
```
does this double initialization control matter for this module? There is no harm in initializing twice. Furthermore - this protection in module creates a false perception that nobody will call `Initialize` twice. In reality it may easily happen with re-creating of entire `TelemetryConfiguration`

d3r3kk replied:
I don't need to protect it, no. I do need to know when it gets re-initialized but that's all.

SergeyKanzhelev commented on Src/WindowsServer/WindowsServer.Shared/AppServicesHeartbeatTelemetryModule.cs:
```diff
@@ -0,0 +1,150 @@
+            new KeyValuePair<string, string>("appSrv_wsHost", "WEBSITE_HOSTNAME")
+        };
+
+        private static AppServicesHeartbeatTelemetryModule instance;
```
why singleton?

d3r3kk replied:
To allow access to the 'Update' method. I don't have a better mechanism at this moment, but I think we can define a better one given some time.

SergeyKanzhelev commented on Src/WindowsServer/WindowsServer.Shared/AppServicesHeartbeatTelemetryModule.cs:
```diff
@@ -0,0 +1,150 @@
+        }
+
+        /// <summary>
+        /// Gets a value to provide internal access to the only instance of this class that *should* be available.
```
I don't understand why it is a singleton. Can you provide more details

d3r3kk replied:
We need a way to allow for the update of the heartbeat values associated to AppServices. I made this a singleton to allow other modules to signal this module to update the values. Perhaps there is a better mechanism for this, but I am not aware of one at this time. (I have thoughts on how to improve it in the future though, but it might take some time).

SergeyKanzhelev commented on Src/WindowsServer/WindowsServer.Shared/Implementation/WindowsServerEventSource.cs:
```diff
@@ -326,6 +326,29 @@ public void AzureInstanceMetadataFailureSettingDefaultPayload(string exception,
+
+        [Event(34,
+            Message = "Could not gain access to the Heartbeat Manager instance during initialization. Exception raised: {0}",
+            Level = EventLevel.Verbose)]
```
Make it warning at least. This error is very unlikely to happen. It indicates that something is broken in SDK. So we want to have it at least at warning level.

d3r3kk replied:
Agreed. I've gone over the other calls and altered their levels (including some I had set previously for the Azure IMS data as well).

SergeyKanzhelev commented on Src/WindowsServer/WindowsServer.Shared/AppServicesHeartbeatTelemetryModule.cs:
```diff
@@ -0,0 +1,150 @@
+                        string hbeatKey = kvp.Key.ToString();
+                        if (isUpdateOperation)
+                        {
+                            hbeatManager.SetHeartbeatProperty(hbeatKey, hbeatValue);
```
Would you expect `hbeatManager` to be `null` ever?

d3r3kk replied:
Yes, I've added an early exit to counter this possibility.

SergeyKanzhelev commented on Src/WindowsServer/WindowsServer.Shared/AppServicesHeartbeatTelemetryModule.cs:
```diff
@@ -0,0 +1,150 @@
+                WindowsServerEventSource.Log.HeartbeatManagerAccessFailure(hearbeatManagerAccessException.Message);
+            }
+
+            return hbeatManager;
```
You do not check for `null` here. It may be a good idea though

d3r3kk replied:
Agreed. Done.

SergeyKanzhelev commented on Src/WindowsServer/WindowsServer.Shared/AppServicesHeartbeatTelemetryModule.cs:
```diff
@@ -0,0 +1,150 @@
+                try
+                {
+                    // get the variable, then expand it (otherwise we get the name we queried for in the value)
+                    string hbeatValue = Environment.GetEnvironmentVariable(kvp.Value);
```
this may throw security exception when it's not full trust appdomain. It is good idea to have a global `try{}catch` in the `Initialize`

d3r3kk replied:
Agreed. Done.

SergeyKanzhelev commented:
I don't see how you are addressing synchronization with telemetry initializer and slot swap issue

d3r3kk commented on Src/WindowsServer/WindowsServer.Shared/AzureWebAppRoleEnvironmentTelemetryInitializer.cs:
```diff
@@ -31,21 +31,35 @@ public AzureWebAppRoleEnvironmentTelemetryInitializer()
+            else if (!nodeName.Equals(this.lastNodeValue, StringComparison.Ordinal))
+            {
+                // if the AppServices heartbeat telemetry module exists, signal it to update the values in heartbeat
+                AppServicesHeartbeatTelemetryModule.Instance?.UpdateHeartbeatWithAppServiceEnvVarValues();
```
@SergeyKanzhelev, here is where I signal the `AppServicesHeartbeatTelemetryModule` to update the values in the heartbeat properties. This is the reason I chose to go with the Singleton pattern. Perhaps there is a better way to go about this?

d3r3kk commented:
@SergeyKanzhelev ping?

d3r3kk commented:
@SergeyKanzhelev @cijothomas updated PR with tests, can be merged without any need for follow up PR.

- NOTE: I've added logic in this PR to update environment variables from the Initializer. That will work, but there is a danger that we would potentially miss updated environment variables (the value is potentially cached, please let me know if that is the case when using `InternalContext` as we do) and there might be perf issues with the way I propose here. Please see PR #870 for an alternate that I'm a bit happier with.

cijothomas commented on Src/WindowsServer/WindowsServer.Shared.Tests/AppServicesHeartbeatTests.cs:
```diff
@@ -0,0 +1,124 @@
+                envVars.Add(kvp.Value, Environment.GetEnvironmentVariable(kvp.Value));
+                if (string.IsNullOrEmpty(envVars[kvp.Value]))
+                {
+                    Environment.SetEnvironmentVariable(kvp.Value, kvp.Key);
```
The effect is setting the ENV variable persists even after the test? This may have unintended consequences in the next test. 
If modifying any ENV, make sure to set it back to original after the test

d3r3kk replied:
Agreed. Also, in PR #870 see that I've added a ClassInitializer/ClassCleanup method to handle environment changes.

d3r3kk replied:
Merged #870 here where I fixed this issue by using unique env vars in place of the standard during testing. Note that I left the original testing for the TelemetryInitializer intact (although it really should be updated to do the same).

cijothomas commented on Src/WindowsServer/WindowsServer.Shared.Tests/WindowsServer.Shared.Tests.projitems:
```diff
@@ -9,6 +9,7 @@
     <Import_RootNamespace>Microsoft.ApplicationInsights.WindowsServer</Import_RootNamespace>
   </PropertyGroup>
   <ItemGroup>
+    <Compile Include="$(MSBuildThisFileDirectory)AppServicesHeartbeatTests.cs" />
```
AppServicesHeartbeatTelemetryModuleTests might be better name.

d3r3kk replied:
Ok.

cijothomas commented on Src/WindowsServer/WindowsServer.Shared/AppServicesHeartbeatTelemetryModule.cs:
```diff
@@ -0,0 +1,181 @@
+                {
+                    try
+                    {
+                        // get the variable, then expand it (otherwise we get the name we queried for in the value)
```
I am not sure i understood this. Why not just read ENV variable like done everywhere else? 
https://github.com/Microsoft/ApplicationInsights-dotnet-server/blob/develop/Src/WindowsServer/WindowsServer.Shared/AzureWebAppRoleEnvironmentTelemetryInitializer.cs#L65

d3r3kk replied:
Agreed. There doesn't ever seem to be any need to handle environment variable expansion for this value. PR #870 does not attempt expansion.

Dmitry-Matveev replied:
`ExpandEnvironmentVariable` will make this method slower, let's skip if unnecessary.

cijothomas commented on Src/WindowsServer/WindowsServer.Shared/AppServicesHeartbeatTelemetryModule.cs:
```diff
@@ -0,0 +1,181 @@
+        /// </summary>
+        internal static readonly KeyValuePair<string, string>[] WebHeartbeatPropertyNameEnvVarMap = new KeyValuePair<string, string>[]
+        {
+            new KeyValuePair<string, string>("appSrv_SiteName", "WEBSITE_SITE_NAME"),
```
Aren't these values appended to all telemetry items by the corresponding TelemetryInitializers? Why do we send these values (duplicated) into HeartBeat again?

d3r3kk replied:
Heartbeat and telemetry initializers are two separate solutions. I am not 100% aware of all the downstream consumers of the TelemetryInitializers, but I am keenly aware of the heartbeat ones (app 2 infra). I want the app 2 infra feature to only look in 1 single place for the values it requires. I believe the duplication caused by heartbeats (default 15-minute cadence) is minimal enough to support the end purpose.

cijothomas commented on Src/WindowsServer/WindowsServer.Shared/AppServicesHeartbeatTelemetryModule.cs:
```diff
@@ -0,0 +1,181 @@
+        /// Environment variables we use in our heartbeat payload.
+        /// </summary>
+        /// <returns>A value indicating whether or not an update to the environment variables occurred.</returns>
+        public bool UpdateHeartbeatWithAppServiceEnvVarValues()
```
Instead of a TelemetryInitializers telling this module to update, can this module update itself (re-read ENV vars) every 'n' minute? And maybe inform the TelemetryInitializer that ENV values changed so that it'll also pick new values.

d3r3kk replied:
Agreed. I was thinking that a subscriber model would work really nicely actually - that way the heartbeat module and the initializer module could simply 'subscribe' to the environment variable monitor (see PR #870, that's the terminology I use there) and get updated only as necessary. 

For this PR, I wanted to keep it as concise as possible to get the job done. I would certainly be open to creating such a subscriber in our next sprint.

d3r3kk replied:
After discussions with Dmitry I will create an alternate PR to this one that contains a subscriber model solution. Although due to time constraints we still may opt for this solution.

cijothomas commented on Src/WindowsServer/WindowsServer.Shared/AppServicesHeartbeatTelemetryModule.cs:
```diff
@@ -0,0 +1,181 @@
+        /// <param name="configuration">Unused parameter.</param>
+        public void Initialize(TelemetryConfiguration configuration)
+        {
+            this.isInitialized = this.UpdateHeartbeatWithAppServiceEnvVarValues();
```
this.isInitialized -- Where is this used?

d3r3kk replied:
Line 94 - in the call to AddAppServiceEnvironmentVariablesToHeartbeat. The first time (isInitialized=false) we use AddHeartbeatProperty, each subsequent time we use SetHeartbeatProperty.

cijothomas commented:
Need to ensure the tests are run in CI builds. I searched for newly added tests but couldn't find them in test results for this CI build.

d3r3kk commented on Src/WindowsServer/WindowsServer.Shared/AzureWebAppRoleEnvironmentTelemetryInitializer.cs:
```diff
@@ -31,21 +31,35 @@ public AzureWebAppRoleEnvironmentTelemetryInitializer()
+            {
+                this.lastNodeValue = nodeName;
+            }
+            else if (!nodeName.Equals(this.lastNodeValue, StringComparison.Ordinal))
```
Agreed. Please see PR #870 where I fix this issue you call out here, and where I am adding some protection for performance as well. (My intention is to merge that PR into this one, as I prefer that solution).

d3r3kk replied:
Merged #870 here. Fixed.

d3r3kk commented:
@cijothomas 
>Need to ensure the tests are run in CI builds. I searched for newly added tests but couldn't find them in test results for this CI build.

There was some iteration on builds as I added the tests, perhaps some of them got added as 'New Tests' in prior builds from the one you checked (I looked, and in the .1 build for that day there was 1 newly added test).

**HOWEVER**

Tests in the `WindowsServer.Test.Shared` project where not being added into the `WindowsServer.Net45.Tests` project *at all*. (`WindowsServer.NetCore.Tests` did load them, so they were still covered, just not fully). I have corrected this oversight in this PR as well.

d3r3kk commented:
This PR is replaced by PR #873.
